contracts/sol6/reserves/KyberFprReserveV2.sol

```diff
@@ -0,0 +1,349 @@
+        IERC20 srcToken,
+        uint256 srcAmount,
+        IERC20 destToken,
+        address payable destAddress,
+        uint256 conversionRate,
+        bool validate
+    ) internal {
+        // can skip validation if done at kyber network level
+        if (validate) {
+            require(conversionRate > 0, "rate is 0");
+            if (srcToken == ETH_TOKEN_ADDRESS) require(msg.value == srcAmount, "wrong msg value");
+            else require(msg.value == 0, "bad msg value");
+        }
+
+        uint256 destAmount = getDestQty(srcToken, destToken, srcAmount, conversionRate);
+        // sanity check
+        require(destAmount > 0, "dest amount is 0");
+
+        // add to imbalance
+        IERC20 token;
+        int256 tradeAmount;
+        if (srcToken == ETH_TOKEN_ADDRESS) {
+            tradeAmount = int256(destAmount);
+            token = destToken;
+        } else {
+            tradeAmount = -1 * int256(srcAmount);
+            token = srcToken;
+        }
+
+        conversionRatesContract.recordImbalance(token, tradeAmount, 0, block.number);
+
+        // collect src tokens, transfer to token wallet if needed
+        address srcTokenWallet = getTokenWallet(srcToken);
+        if (srcToken == ETH_TOKEN_ADDRESS) {
+            if (srcTokenWallet != address(this)) {
+                // reserve is using weth instead of eth
+                // convert eth to weth and send to weth's token wallet
+                weth.deposit{value: msg.value}();
+                IERC20(weth).safeTransfer(srcTokenWallet, msg.value);
+            }
+        } else {
+            srcToken.safeTransferFrom(msg.sender, srcTokenWallet, srcAmount);
+        }
+
+        address destTokenWallet = getTokenWallet(destToken);
+        // send dest tokens
+        if (destToken == ETH_TOKEN_ADDRESS) {
+            if (destTokenWallet != address(this)) {
+                // transfer weth from weth's token wallet to this address
+                // then unwrap weth to eth to this contract address
+                IERC20(weth).safeTransferFrom(destTokenWallet, address(this), destAmount);
+                weth.withdraw(destAmount);
+            }
+            // transfer eth to dest address
+            (bool success, ) = destAddress.call{value: destAmount}("");
+            require(success, "transfer back eth failed");
+        } else {
+            if (destTokenWallet == address(this)) {
+                destToken.safeTransfer(destAddress, destAmount);
+            } else {
+                destToken.safeTransferFrom(destTokenWallet, destAddress, destAmount);
+            }
+        }
+
+        emit TradeExecute(msg.sender, srcToken, srcAmount, destToken, destAmount, destAddress);
+    }
+
+    /// @dev return wallet that holds the token
+    ///      if token is ETH, check tokenWallet of WETH instead
+    ///      if wallet is 0x0, consider as this reserve address
+    function getTokenWallet(IERC20 token) internal view returns (address wallet) {
```

Comment:
we should consider making this public?
might help reserve manager.

Reply:
+1, I would say it should be public. Not just reserve manager, but for debugging purposes